tidy3d/components/data/monitor_data.py

```diff
@@ -996,6 +1009,111 @@ def to_source(
             field_dataset=dataset, source_time=source_time, center=center, size=size, **kwargs
         )
 
+    def make_adjoint_sources(
+        self, dataset_names: list[str]
+    ) -> List[Union[CustomCurrentSource, PointDipole]]:
+        """Converts a :class:`.FieldData` to a list of adjoint current or point sources."""
+
+        if np.allclose(np.array(self.monitor.size), np.zeros(3)):
+            return self.to_adjoint_point_sources()
+
+        return self.to_adjoint_field_sources()
+
+    def to_adjoint_point_sources(self) -> List[PointDipole]:
+        """Create adjoint point dipole source if this field data contains one item."""
+
+        sources = []
+
+        for polarization, field_component in self.field_components.items():
+            if field_component is None:
+                continue
+
+            for freq0 in field_component.coords["f"]:
+                omega0 = 2 * np.pi * freq0
+                scaling_factor = 1 / (MU_0 * omega0)
+
+                forward_amp = complex(np.squeeze(field_component.sel(f=freq0).values))
+
+                adj_phase = 3 * np.pi / 2 + np.angle(forward_amp)
+
+                adj_amp = scaling_factor * forward_amp
+
+                src_adj = PointDipole(
+                    center=self.monitor.center,
+                    polarization=polarization,
+                    source_time=GaussianPulse(
+                        freq0=freq0,
+                        fwidth=freq0 / 10,  # TODO: how to set this properly?
+                        amplitude=abs(adj_amp),
+                        phase=adj_phase,
+                    ),
+                    interpolate=True,
+                )
+
+                sources.append(src_adj)
+
+        return sources
+
+    def to_adjoint_field_sources(self) -> List[CustomCurrentSource]:
+        """Create adjoint custom field sources if this field data has some dimensionality."""
+
+        sources = []
+
+        # Define source geometry based on coordinates in the data
+        data_mins = []
+        data_maxs = []
+
+        def shift_value(coords) -> float:
+            """How much to shift the geometry by along a dimension (only if > 1D)."""
+            return 1e-5 if len(coords) > 1 else 0
```

Comment:
Should this `1e-5` be set somewhere else maybe as a global constant?

Reply:
yea maybe.. I dont remember where this came from exactly. @momchil-flex do you remember? I guess we just wanted to make sure that the source was always to the same side of the coordinate? any updates since then in the meshing so I dont need to worry about this as much?

Reply:
but I can move it to a global constant either way.

Reply:
I don't really remember either... We should have left more comments :)